loader/lightning.go

```diff
@@ -0,0 +1,318 @@
+	lightningCfg.TiDB.StatusPort = cfg.TiDB.StatusPort
+	lightningCfg.TiDB.PdAddr = cfg.TiDB.PdAddr
+	lightningCfg.TikvImporter.Backend = cfg.TiDB.Backend
+	if cfg.TiDB.Backend == lcfg.BackendLocal {
+		lightningCfg.TikvImporter.SortedKVDir = cfg.Dir
+	}
+	lightningCfg.Mydumper.SourceDir = cfg.Dir
+	lightningCfg.App.Config.File = cfg.LogFile
+	lightningCfg.App.Config.Level = cfg.LogLevel
+	return lightningCfg
+}
+
+// Type implements Unit.Type.
+func (l *LightningLoader) Type() pb.UnitType {
+	return pb.UnitType_Load
+}
+
+// Init initializes loader for a load task, but not start Process.
+// if fail, it should not call l.Close.
+func (l *LightningLoader) Init(ctx context.Context) (err error) {
+	tctx := tcontext.NewContext(ctx, l.logger)
+	checkpoint, err := newRemoteCheckPoint(tctx, l.cfg, l.checkpointID())
+	failpoint.Inject("ignoreLoadCheckpointErr", func(_ failpoint.Value) {
+		l.logger.Info("", zap.String("failpoint", "ignoreLoadCheckpointErr"))
+		err = nil
+	})
+	l.checkPoint = checkpoint
+	l.toDB, l.toDBConns, err = createConns(tctx, l.cfg, 1)
+	return err
+}
+
+func (l *LightningLoader) restore(ctx context.Context) error {
+	tctx := tcontext.NewContext(ctx, l.logger)
+	if err := l.putLoadTask(); err != nil {
+		return err
+	}
+	if err := l.checkPoint.Init(tctx, lightningCheckpointFile, 1); err != nil {
+		return err
+	}
+	if err := l.checkPoint.Load(tctx); err != nil {
+		return err
+	}
+	db2Tables := make(map[string]Tables2DataFiles)
+	tables := make(Tables2DataFiles)
+	files := make(DataFiles, 0, 1)
+	files = append(files, lightningCheckpointFile)
+	tables[lightningCheckpointTable] = files
+	db2Tables[lightningCheckpointDB] = tables
+	var err error
+	if err = l.checkPoint.CalcProgress(db2Tables); err != nil {
+		return err
+	}
+	if !l.checkPoint.IsTableFinished(lightningCheckpointDB, lightningCheckpointTable) {
+		cfg := lcfg.NewConfig()
+		if err = cfg.LoadFromGlobal(l.lightningConfig); err != nil {
+			return err
+		}
+		cfg.Routes = l.cfg.RouteRules
+		if err = cfg.Adjust(ctx); err != nil {
+			return err
+		}
+		taskCtx, cancel := context.WithCancel(ctx)
+		l.cancel = cancel
+		err = l.core.RunOnce(taskCtx, cfg, nil)
+		if err == nil {
+			l.finish.Store(true)
+			offsetSQL := l.checkPoint.GenSQL(lightningCheckpointFile, 1)
+			err = l.toDBConns[0].executeSQL(tctx, []string{offsetSQL})
+			_ = l.checkPoint.UpdateOffset(lightningCheckpointFile, 1)
+		}
+	} else {
+		l.finish.Store(true)
+	}
+	if l.cfg.Mode == config.ModeFull {
+		_ = l.delLoadTask()
+	}
+	if l.cfg.CleanDumpFile {
+		l.cleanDumpFiles()
+	}
+	return err
+}
+
+// Process implements Unit.Process.
+func (l *LightningLoader) Process(ctx context.Context, pr chan pb.ProcessResult) {
+	l.logger.Info("lightning load start")
+	errs := make([]*pb.ProcessError, 0, 1)
+	if err := l.restore(ctx); err != nil {
+		errs = append(errs, unit.NewProcessError(err))
+	}
+	l.logger.Info("lightning load end")
+	pr <- pb.ProcessResult{
+		IsCanceled: false,
```

Comment:
this field is used to distinguish user canceling context, in current implementation it should be set to `true` if context is canceled (seems we can move this logic to caller 🤔 )

https://github.com/pingcap/dm/blob/6ad2f890822495986e7d98a55d5202b72a8442b9/loader/loader.go#L616-L628